src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -676,37 +676,98 @@ internal static ulong GetSegmentSize()
+        /// </summary>
+        /// <typeparam name="T">Specifies the type of the array element.</typeparam>
+        /// <param name="length">Specifies the length of the array.</param>
+        /// <param name="pinned">Specifies whether the allocated array must be pinned.</param>
+        /// <remarks>
+        /// If pinned is set to true, <typeparamref name="T"/> must not be a reference type or a type that contains object references.
+        /// </remarks>
         [MethodImpl(MethodImplOptions.AggressiveInlining)] // forced to ensure no perf drop for small memory buffers (hot path)
-        internal static unsafe T[] AllocateUninitializedArray<T>(int length)
+        public static unsafe T[] AllocateUninitializedArray<T>(int length, bool pinned = false)
         {
-            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            if (!pinned)
             {
-                return new T[length];
-            }
+                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+                {
+                    return new T[length];
+                }
 
-            // for debug builds we always want to call AllocateNewArray to detect AllocateNewArray bugs
+                // for debug builds we always want to call AllocateNewArray to detect AllocateNewArray bugs
 #if !DEBUG
-            // small arrays are allocated using `new[]` as that is generally faster.
-            if (length < 2048 / Unsafe.SizeOf<T>())
+                // small arrays are allocated using `new[]` as that is generally faster.
+                if (length < 2048 / Unsafe.SizeOf<T>())
+                {
+                    return new T[length];
+                }
+#endif
+            }
+            else if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
-                return new T[length];
+                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
             }
-#endif
+
             // kept outside of the small arrays hot path to have inlining without big size growth
-            return AllocateNewUninitializedArray(length);
+            return AllocateNewUninitializedArray(length, pinned);
 
-            T[] AllocateNewUninitializedArray(int length)
+            static T[] AllocateNewUninitializedArray(int length, bool pinned)
             {
+                GC_ALLOC_FLAGS flags = GC_ALLOC_FLAGS.GC_ALLOC_ZEROING_OPTIONAL;
+                if (pinned)
+                    flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
+
                 if (length < 0)
                     throw new OverflowException();
 
                 T[] array = null;
-                RuntimeImports.RhAllocateUninitializedArray(EETypePtr.EETypePtrOf<T[]>().RawValue, (uint)length, Unsafe.AsPointer(ref array));
+                RuntimeImports.RhAllocateUninitializedArray(EETypePtr.EETypePtrOf<T[]>().RawValue, (uint)length, (uint)flags, Unsafe.AsPointer(ref array));
```

Comment:
Nit: rename to `RhAllocateNewArray` since the uninitialized part is now optional?